File: home/dot_local/share/lazyvim/lua/lazyvim_capabilities/registry.lua
local definitions = {
	require("lazyvim_capabilities.capabilities.foundation"),
	require("lazyvim_capabilities.capabilities.fonts"),
	require("lazyvim_capabilities.capabilities.node"),
	require("lazyvim_capabilities.capabilities.nvim"),
	require("lazyvim_capabilities.capabilities.tmux"),
	require("lazyvim_capabilities.capabilities.languages.go"),
	require("lazyvim_capabilities.capabilities.languages.typescript"),
	require("lazyvim_capabilities.capabilities.languages.standard"),
}

local M = {}

function M.create(context)
	local capabilities, enabled = {}, {}
	for _, capability in ipairs(definitions) do
		assert(not capabilities[capability.id], "duplicate capability: " .. capability.id)
		capabilities[capability.id] = capability
		if capability.supports(context) then
			enabled[capability.id] = true
		end
	end
	for _, capability in ipairs(definitions) do
		for _, dependency in ipairs(capability.requires) do
			assert(capabilities[dependency], capability.id .. " requires unknown capability " .. dependency)
		end
	end

	local ordered, visiting, visited = {}, {}, {}
	local function visit(id)
		if not enabled[id] or visited[id] then
			return
		end
		assert(not visiting[id], "capability dependency cycle at " .. id)
		visiting[id] = true
		for _, dependency in ipairs(capabilities[id].requires) do
			assert(enabled[dependency], id .. " requires unsupported capability " .. dependency)
			visit(dependency)
		end
		visiting[id], visited[id] = nil, true
		table.insert(ordered, capabilities[id])
	end
	for _, capability in ipairs(definitions) do
		visit(capability.id)
	end

	local registry = { ordered = ordered }
	function registry:enhancements_for(target)
		local result = {}
		for _, capability in ipairs(self.ordered) do
			vim.list_extend(result, capability.enhancements[target] or {})
		end
		return result
	end
	function registry:run(lifecycle)
		for _, capability in ipairs(self.ordered) do
			if capability[lifecycle] then
				print(("\n==> %s capability: %s"):format(lifecycle, capability.id))
				local hook_context =
					vim.tbl_extend("force", context, { enhancements = self:enhancements_for(capability.id) })
				capability[lifecycle](hook_context)
			end
		end
	end
	return registry
end

return M

File: home/dot_local/share/lazyvim/capabilities/registry.mjs
import foundation from "./foundation.mjs";
import fonts from "./fonts.mjs";
import go from "./languages/go.mjs";
import standardLanguages from "./languages/standard.mjs";
import typescript from "./languages/typescript.mjs";
import node from "./node.mjs";
import nvim from "./nvim.mjs";
import tmux from "./tmux.mjs";

const definitions = [
  foundation,
  fonts,
  node,
  nvim,
  tmux,
  go,
  typescript,
  standardLanguages,
];

export function createCapabilityRegistry(context) {
  const capabilities = new Map(definitions.map((item) => [item.id, item]));
  if (capabilities.size !== definitions.length) {
    throw new Error("Capability ids must be unique");
  }
  for (const capability of capabilities.values()) {
    for (const dependency of capability.requires) {
      if (!capabilities.has(dependency)) {
        throw new Error(
          `${capability.id} requires unknown capability ${dependency}`,
        );
      }
    }
  }

  const enabled = new Set(
    definitions.filter((item) => item.supports(context)).map((item) => item.id),
  );
  const visiting = new Set();
  const visited = new Set();
  const ordered = [];
  function visit(id) {
    if (!enabled.has(id) || visited.has(id)) return;
    if (visiting.has(id))
      throw new Error(`Capability dependency cycle at ${id}`);
    visiting.add(id);
    for (const dependency of capabilities.get(id).requires) {
      if (!enabled.has(dependency)) {
        throw new Error(`${id} requires unsupported capability ${dependency}`);
      }
      visit(dependency);
    }
    visiting.delete(id);
    visited.add(id);
    ordered.push(capabilities.get(id));
  }
  for (const id of enabled) visit(id);

  function enhancementsFor(target) {
    return ordered.flatMap(
      (capability) => capability.enhancements[target] || [],
    );
  }

  async function run(lifecycle) {
    for (const capability of ordered) {
      const hook = capability[lifecycle];
      if (!hook) continue;
      console.log(`\n==> ${lifecycle} capability: ${capability.id}`);
      await hook({ ...context, enhancements: enhancementsFor(capability.id) });
    }
  }

  return Object.freeze({ enhancementsFor, ordered, run });
}
